# 설치

pip install torch

pip install mxnet

pip install gluonnlp==0.8.0

0.9.1은 이유를 모르겠는데 설치가 안되더라고요

pip install sentencepiece

pip install onnxruntime

pip install transformers

pip install kobert-transformers

### kobert-transformers 사이트

https://pypi.org/project/kobert-transformers/

https://github.com/monologg/KoBERT-Transformers

In [1]:
import torch
import mxnet
import gluonnlp
import sentencepiece
import onnxruntime
import transformers
import kobert_transformers

C:\Users\golds\AppData\Roaming\Python\Python37\site-packages\mxnet\optimizer\optimizer.py:167: UserWarning: WARNING: New optimizer gluonnlp.optimizer.lamb.LAMB is overriding existing optimizer mxnet.optimizer.optimizer.LAMB
  Optimizer.opt_registry[name].__name__))


In [2]:
print('torch : ', torch.__version__)
print('mxnet : ', mxnet.__version__)
print('gluonnlp : ', gluonnlp.__version__)
# print('sentencepiece : ', sentencepiece.__version__)
print('onnxruntime : ', onnxruntime.__version__) 
print('transformers : ', transformers.__version__)
print('kobert_transformers : ', kobert_transformers.__version__) 

torch :  1.6.0
mxnet :  1.6.0
gluonnlp :  0.8.0
onnxruntime :  1.4.0
transformers :  3.0.2
kobert_transformers :  0.4.1


In [3]:
import numpy as np
import pandas as pd
import sqlite3

# 예시로 사용할 데이터 불러오기

In [4]:
conn = sqlite3.connect('./news_cleaned.db')
cur = conn.cursor()

In [5]:
query = cur.execute('SELECT * FROM daumnews LIMIT 0, 1000')
cols = [column[0] for column in query.description]

In [6]:
daum = pd.DataFrame.from_records(data=query.fetchall(), columns=cols)

In [7]:
cur.close()
conn.close()

In [8]:
print(daum.body[0])

"벌금과 단속 걱정하던 아들, 장사 준비하던 현수막으로 목 메 숨져"결혼까지 미루고 식당 영업재개할 생각에 청소까지 열심히 했는데···"

 "아버지 아들인게 정말 행복했습니다. 이 가게가 잘 될 수 있도록 수호신이 될게요"
26살 황승우씨는 지난 30일 오전 경기 남양주시 조안면 막국수집 주방에서 목을 메 숨진 채 발견됐다. 이곳은 아버지 황선남(65)씨와 지난 2015년 희망을 품고 문을 연 가게였다.
건물주의 운전기사로 시작해 세를 얻어 희망을 품고 시작한 막국수 집은 인상이 서글서글한 부자의 웃음과 열의로 그나마 장사가 되는 듯 했다.
그러나 이곳은 팔당상수원 환경정비구역으로 남양주시와 의정부지검이 지난해 부터 단속을 벌여 음식점을 운영하던 7명이 구속됐다.
황씨가 운영하던 막국수 집도 단속을 피할 수 없었다. 지난해 12월 단속을 더이상 견딜 수 없었던 이들은 결국 음식점 문을 닫았다.
문제는 수입은 없는데 검찰이 내린 벌금 3000만원과 남양주시의 이행강제금 3690만원을 감당하기에는 너무나 힘에 부쳤다. 주변 가게에서 아르바이트를 해도 생활비는 커녕 벌금과 과태료도 낼 수 없었다.
결국 승우씨는 주변 지인과 카드빚을 내 휴가철 관광객들을 대상으로 소시지와 커피 등을 판매하는 노점상을 하기로 했다. 교제하던 여자친구와는 결혼도 미뤘다.
하지만 이마저도 한강유역환경청이 휴가철 수질을 오염시킬 우려가 있다며 지난달 집중단속을 하면서 승우씨의 노점상도 손해만 본 채 접게 됐다. 승우씨는 지인들 뿐 아니라 금융권에서도 신용불량자의 낙인이 찍힌 채 절망에 빠졌다.
이런 가운데 환경청에서 일부 원주민들에 대해서는 음식점 운영을 할 수 있도록 제도를 개선한다는 소식이 전해지면서 한줄기 희망이 생기는 듯 했다. 승우씨는 다시 장사를 할 수도 있다는 생각에 문을 닫았던 음식점의 주방부터 청소를 시작했다.
하지만 환경청이 주민들의 의견을 묵살하면서 이같은 희망도 오래가지 못했다. 승우씨는 청소를 위해 음식점을 찾았다가 이같은 소식을 접했다.

음식점 앞에는 독촉장이 

In [9]:
print(daum.summary[0])

26살 황승우씨는 지난 30일 오전 경기 남양주시 조안면 막국수집 주방에서 목을 메 숨진 채 발견됐다.
이곳은 아버지 황선남(65)씨와 지난 2015년 희망을 품고 문을 연 가게였다.
황씨가 운영하던 막국수 집도 단속을 피할 수 없었다.
하지만 이마저도 한강유역환경청이 휴가철 수질을 오염시킬 우려가 있다며 지난달 집중단속을 하면서 승우씨의 노점상도 손해만 본 채 접게 됐다.


# Tokenizing

### sktbrain-kobert 예시 코드

In [10]:
from gluonnlp.data import SentencepieceTokenizer
from kobert.utils import get_tokenizer

In [11]:
tok_path = get_tokenizer()
sp  = SentencepieceTokenizer(tok_path)

using cached model


In [12]:
example_body_token = sp(daum.body[0])
example_body_token[:20]

['▁"',
 '벌',
 '금',
 '과',
 '▁단속',
 '▁걱정',
 '하던',
 '▁아들',
 ',',
 '▁장',
 '사',
 '▁준비',
 '하던',
 '▁현',
 '수',
 '막',
 '으로',
 '▁목',
 '▁메',
 '▁숨']

In [13]:
example_summary_token = sp(daum.summary[0])
example_summary_token[:20]

['▁26',
 '살',
 '▁황',
 '승',
 '우',
 '씨는',
 '▁지난',
 '▁30',
 '일',
 '▁오전',
 '▁경기',
 '▁남',
 '양',
 '주',
 '시',
 '▁조',
 '안',
 '면',
 '▁막',
 '국']

어떻게 단어를 id로 바꾸는지 모르겠음...!

### kobert_transformers 예시 코드

In [14]:
from kobert_transformers import get_tokenizer

In [15]:
tokenizer = get_tokenizer()

In [16]:
example_body_token = tokenizer.tokenize(daum.body[0])
example_body_token[:20]

['▁"',
 '벌',
 '금',
 '과',
 '▁단속',
 '▁걱정',
 '하던',
 '▁아들',
 ',',
 '▁장',
 '사',
 '▁준비',
 '하던',
 '▁현',
 '수',
 '막',
 '으로',
 '▁목',
 '▁메',
 '▁숨']

In [17]:
example_summary_token = tokenizer.tokenize(daum.summary[0])
example_summary_token[:20]

['▁26',
 '살',
 '▁황',
 '승',
 '우',
 '씨는',
 '▁지난',
 '▁30',
 '일',
 '▁오전',
 '▁경기',
 '▁남',
 '양',
 '주',
 '시',
 '▁조',
 '안',
 '면',
 '▁막',
 '국']

In [18]:
example_body_token_id = tokenizer.convert_tokens_to_ids(example_body_token)
example_body_token_id[:20]

[518,
 6332,
 5550,
 5468,
 1590,
 880,
 7802,
 3107,
 46,
 3954,
 6493,
 4249,
 7802,
 5049,
 6629,
 6149,
 7078,
 2068,
 2016,
 2919]

In [19]:
example_summary_token_id = tokenizer.convert_tokens_to_ids(example_summary_token)
example_summary_token_id[:20]

[584,
 6519,
 5149,
 6703,
 7005,
 6787,
 4304,
 592,
 7126,
 3431,
 956,
 1409,
 6853,
 7276,
 6705,
 4162,
 6812,
 6198,
 1927,
 5503]

# Embedding

### sktbrain-kobert 예시 코드

from kobert.pytorch_kobert import get_pytorch_kobert_model

model, vocab  = get_pytorch_kobert_model()

### kobert_transformers 예시 코드

In [20]:
from kobert_transformers import get_kobert_model

In [21]:
model = get_kobert_model()
model.eval()

BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(8002, 768, padding_idx=1)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0): BertLayer(
        (attention): BertAttention(
          (self): BertSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )

In [22]:
example_body_tensor = torch.LongTensor([example_body_token_id])
example_summary_tensor = torch.LongTensor([example_summary_token_id])

In [23]:
example_body_tensor.shape

torch.Size([1, 883])

In [24]:
try:
    body_sequence_output, body_pooled_output = model(example_body_tensor)
except IndexError as e:
    print(e)

index out of range in self


embedding vector의 차원이 768이니까 input의 길이가 768보다 커서 index error가 뜨는것인가...?

단어 개수가 768개보다 작은거만 input으로 넣을 수 있다는게 말이 안되지 않나...?

In [25]:
summary_sequence_output, summary_pooled_output = model(example_summary_tensor)

In [26]:
example_summary_tensor.shape

torch.Size([1, 116])

In [27]:
summary_sequence_output.shape, summary_pooled_output.shape

(torch.Size([1, 116, 768]), torch.Size([1, 768]))

1개의 sample

단어수가 116개

embedding vector의 차원이 768